Enable position independent code, more miscellaneous fixes

MaartenBent opened the pull request:
Fix typos and warnings.
Improvements to CMake. Using build-in libraries also requires some small changes to expat and tiff, I'll add them later.
Fix building with stl enabled and VS2019 16.6.
Enable position independent code.

MaartenBent committed:
Fix typos in wxSTC interface

MaartenBent committed:
Make sure generated wxSTC files have LF EOL

Use 'wb' with Python 2 to enforce LF newlines. This does not work with
Python 3, use 'newline':'\n' there (which in turn does not work with 2).

MaartenBent committed:
CMake: allow to use build-in libraries on Linux

Requires updates of expat and tiff submodules.

Most libraries have their own CMakeLists, add a comment with a suggestion to
use this for implementing build-in support.

MaartenBent committed:
CMake: fix name clash between folder and application

The test_gui project copies some resource files into image and drawing folders,
so don't use the same name for the sample applications.

MaartenBent committed:
CMake: use dedicated function to check function availability

MaartenBent committed:
Fix implicit-int-float-conversion warning in bombs demo

MaartenBent committed:
Fix unused-function warning in wxSockAddressImpl code

MaartenBent committed:
Fix unused-variable warning in wxListCtrl code

MaartenBent committed:
CMake: link with javascriptcoregtk-4.0 when building static library

MaartenBent committed:
CMake: use set_target_properties() instead of set_property(TARGET )

vadz commented on include/wx/hashmap.h:
```diff
@@ -38,6 +38,10 @@
 #endif
 
+#if defined(_MSC_VER) && _MSC_VER > 1925
+    #define _WX_NO_STL_WITH_DECL
```
Sorry, I don't understand this: didn't we decide that we should never mark these classes as DLL-exported, for all compilers, as they had only inline methods anyhow?

If I'm missing something and we still need to export them, I think this could be done in a simpler way by replacing
```
CLASSEXP CLASSNAME                                                           
```
line with some
```
WX_HASH_APPLY_EXPORT(CLASSEXP) CLASSNAME
```
and then defining `WX_HASH_APPLY_EXPORT` macro as just doing nothing for MSVC >= 1926.

But, again, it would be best not to export them at all, AFAICS...

MaartenBent replied:
I tried to follow your suggestion in https://github.com/wxWidgets/wxWidgets/pull/1993#issuecomment-663274123
> I think we should leave _WITH_DECL but ignore DECL in them for MSVC (maybe just MSVC 19+). This should be an even more minimal change.

Maybe I understood wrong? I don't mind ignoring it for all. But I'll drop the commit from this PR and create a new one.

MaartenBent replied:
See #2009

vadz replied:
Sorry, I forgot that I wrote "for MSVC". But looking at the code now it seems that it should be fine to do it for everybody. We do have other fully-inline classes that are not DLL-exported already and I don't remember having any problems with them...

Thanks again for working on this and sorry again for confusion/wrong suggestion!

vadz reviewed (commented):
Just FYI: I've only looked at the MSVS commit, as I think it should be merged separately (it doesn't seem to be really related to the rest of the changes and is more important/critical), I'll check the rest of them slightly later.

vadz commented on src/stc/scintilla/include/Scintilla.iface:
```diff
@@ -331,7 +331,7 @@ set void MarkerSetBack=2042(int markerNumber, colour back)
 set void MarkerSetBackSelected=2292(int markerNumber, colour back)
 
-# Enable/disable highlight for current folding bloc (smallest one that contains the caret)
+# Enable/disable highlight for current folding block (smallest one that contains the caret)
```
Should we check this file in `misc/scripts/spellcheck`?

MaartenBent replied:
No, this is a file from Scintilla. I don't think we should spellcheck files from external projects.
You fixed this in `include/stc/stc/h` and `interface/stc/stc.h` when you added spellcheck, but these changes will be overwritten when these files are regenerated.
I think it is better to fix this in the Scintilla source, than to ignore the stc files/lines in spellcheck. It is also fixed in a later version if Scintilla, so it should fix itself once Scintilla get updated.

vadz replied:
> You fixed this in `include/stc/stc/h` and `interface/stc/stc.h` when you added spellcheck,

Oh, indeed, thanks for noticing this, I totally failed to realize that I was editing a generated file, even if I really should have known better (but actually I didn't edit it, the changes were done by codespell itself, with its `-w` option, which is why I didn't even notice which file was being changed).

> I think it is better to fix this in the Scintilla source, than to ignore the stc files/lines in spellcheck. It is also fixed in a later version if Scintilla, so it should fix itself once Scintilla get updated.

Yes, absolutely, thanks again.

vadz commented on src/stc/gen_iface.py:
```diff
@@ -1174,11 +1174,18 @@ def processIface(iface, h_tmplt, cpp_tmplt, ih_tmplt, h_dest, cpp_dest, docstr_d
     # write out destination files
-    open(h_dest, 'w').write(h_text)
-    open(cpp_dest, 'w').write(cpp_text)
+    if sys.version_info[0] == 2:
+        acc = 'wb'
+        eol = {}
+    else:
+        acc = 'wt'
+        eol = {'newline':'\n'}
+
+    open(h_dest, acc, **eol).write(h_text)
+    open(cpp_dest, acc, **eol).write(cpp_text)
```
Just to be sure: is the intention here to always write files using LF EOL under all platforms?

If so, why can't we just always use "wb", even for Python 3?

Sorry if I'm missing something obvious here again, but if using "wb" doesn't work, perhaps it could be briefly explained why it doesn't in a comment?

MaartenBent replied:
Yes, this is to make sure LF EOL is used in the generated files (`stc.h`, `stc.cpp`). On Windows, at least for me, they got replaced with CRLF.

Python 3 requires `'newline':'\n'` to get the correct results. `wb` does not work:  `TypeError: a bytes-like object is required, not 'str'`. It will also work with only `w`, but for consistency I used `wt`.
Python 2 requires `wb` for correct results. It does not like `'newline':'\n'`: `TypeError: 'newline' is an invalid keyword argument for this function`. Using `wt` has no effect.

I'll add a comment to the code.

vadz reviewed (approved):
Thanks, no real comments, just a couple of minor questions.

vadz commented on configure:
```diff
@@ -32895,6 +32926,11 @@ if test "$wxUSE_PROFILE" = "yes" ; then
     PROFILE_FLAGS=" -pg"
 fi
 
+PIC_FLAGS=
+if test "$wxUSE_PIC" = "yes" ; then
+    PIC_FLAGS=" -fpic"
+fi
+
```
We already have `PIC_FLAG` defined in `AC_BAKEFILE_SHARED_LD` from `build/aclocal/bakefile.m4` (which is, wrongly but conveniently for us, called unconditionally, i.e. even when building static libraries), so I think we should use it instead (even if `-fpic` is potentially more efficient). Is there any reason you have added this one or was it just an oversight?

MaartenBent replied:
I didn't know I could use that one. But that file is only included after most of the compiler arguments are set, so I can't add it to `C_AND_CXX_FLAGS` like I do now, but have to do something like
```
if test "$wxUSE_PIC" = "yes" ; then
    CFLAGS=`echo $CFLAGS $PIC_FLAG `
    CXXFLAGS=`echo $CXXFLAGS $PIC_FLAG `
    OBJCFLAGS=`echo $OBJCFLAGS $PIC_FLAG `
    OBJCXXFLAGS=`echo $OBJCXXFLAGS $PIC_FLAG `
fi
```

Should I also add it to `WXCONFIG_CPPFLAGS`?

vadz replied:
No, I don't think this should go into `WXCONFIG_CPPFLAGS` (and it's not a cpp flag anyhow), there doesn't seem to be any good reason to force compiling any code using wx as PIC, if the application wants to use PIC (or, more likely, PIE) flag for its code, it should do it on its own.

vadz commented on configure.in:
```diff
@@ -678,6 +678,7 @@ WX_ARG_ENABLE(extended_rtti, [  --enable-extended_rtti  use extended RTTI (XTI)]
 WX_ARG_DISABLE(optimise,   [  --disable-optimise      compile without optimisations], wxUSE_OPTIMISE)
 
 WX_ARG_ENABLE(profile,       [  --enable-profile        create code with profiling information], wxUSE_PROFILE)
+WX_ARG_DISABLE(pic,          [  --disable-pic           disable position independent code], wxUSE_PIC)
```
```suggestion
WX_ARG_DISABLE(pic,          [  --disable-pic           don't use position independent code when building static libraries (shared libraries always use PIC)], wxUSE_PIC)
```
This might be too long, but I think the current message is a bit misleading...

Alternatively, we could give an error if `--disable-pic` is specified, but `--disable-shared` is not, what do you think?

MaartenBent replied:
I'll update the description.
I can add a warning, like is done for `wxUSE_PLUGINS`.

MaartenBent committed:
Enable position independent code by default

Add a build option (--disable-pic, wxBUILD_PIC=OFF) to disable it.

Note that it is always enabled for shared libraries and static third party libraries.

vadz commented:
FYI: this results in a problem with Qt build, although I'm not sure why.

MaartenBent commented:
Thanks for restarting those builds. I think it fails because Qt build used to add it to `WXCONFIG_CXXFLAGS` as well, which is used when building the tests?
I'll restore that line (https://github.com/wxWidgets/wxWidgets/pull/2008/commits/fd86fbc1fe7a37cd85848c76d33a9d82a4087798#diff-e2d5a00791bce9a01f99bc6fd613a39dL26311).

MaartenBent commented:
Other conclusion. The PIC flags were never set to `SAMPLES_CXXFLAGS` when building shared libraries.

MaartenBent commented:
I think I fixed everything. It is ready to merge.

vadz commented:
Thanks again!
